docs: add blog Database Change Management with Snowflake (#709)
bytebase/bytebase.com: content/blog/database-change-management-with-snowflake.md

---
title: Database Change Management with Snowflake
author: Ningjing
published_at: 2022/12/22 21:15
feature_image: /static/blog/database-change-management-with-snowflake/snowflake-db-management.webp
tags: How-To
description: This tutorial will guide you step-by-step to set up database change management for Snowflake in Bytebase. 
---

This is a series of articles about Database Change Management with Snowflake

- Database Change Management with Snowflake (this one)
- Database Change Management with Snowflake and GitHub

---

> This tutorial will guide you step-by-step to set up database change management for Snowflake in Bytebase. With Bytebase, a team can have a formalized review and rollout process to make Snowflake schema change and data change.

Reviewer: ```suggestion
This tutorial will guide you step-by-step to set up database change management for [Snowflake](https://www.snowflake.com/en/) in Bytebase. With Bytebase, a team can have a formalized review and rollout process to make Snowflake schema and data changes.
```

1. Make sure your docker daemon is running, and start the Bytebase docker container by typing the following command in the terminal.

````bash
docker run --init \
  --name bytebase \
  --restart always \
  --publish 5678:8080 \
  --health-cmd "curl --fail http://localhost:5678/healthz || exit 1" \
  --health-interval 5m \
  --health-timeout 60s \
  --volume ~/.bytebase/data:/var/opt/bytebase \
  bytebase/bytebase:1.9.1 \
  --data /var/opt/bytebase \
  --port 8080 \
````

2. Bytebase is running successfully in Docker, and you can visit it via `localhost:5678`.
![docker](/static/blog/database-change-management-with-snowflake/docker.webp)

3. Visit `localhost:5678` in your browser. Register the first admin account which will be granted `Workspace Owner`.
![bb-register](/static/blog/database-change-management-with-snowflake/bb-register.webp)

## Step 2 - Add Snowflake account in Bytebase
In Bytebase, ​​an **Instance** could be your on-premises MySQL instance, an AWS RDS instance etc, in this tutorial, a Snowflake account.

1. Visit `localhost:5678` and login as Workspace Owner.
![bb-login](/static/blog/database-change-management-with-snowflake/bb-login.webp)

> 2. Follow the onboarding guide if it’s your first time running Bytebase, if not click **Add Instance**.

Reviewer: ```suggestion
2. Follow the onboarding guide if it’s your first time running Bytebase, otherwise, simply click **Add Instance**.
```

![bb-add-instance](/static/blog/database-change-management-with-snowflake/bb-add-instance.webp)

3. Add a Snowflake instance. You need to pay attention to some fields:
![bb-create-instance](/static/blog/database-change-management-with-snowflake/bb-create-instance.webp)

> **Environment**: choose `Test`, if you choose `Prod`, it will wait for approval by default. In this tutorial, we try to keep it simple. However, it’s all configurable.

Reviewer: ```suggestion
**Environment**: choose `Test`, if you choose `Prod`, you will need approval for all future change requests. In this tutorial, let's try to keep it simple. (However, it’s all configurable later.)
```

**Account name**: Go to your Snowflake account, you can find it in the URL, or from the locator field (but lower case).
![sf-locator](/static/blog/database-change-management-with-snowflake/sf-locator.webp)

> **Username and password**: The ones you use to log into Snowflake.

Reviewer: ```suggestion
**Username and password**: The ones you use to log into your Snowflake account.
```

![bb-sf-created](/static/blog/database-change-management-with-snowflake/bb-sf-created.webp)

2. Click **New Project** to create a new project `TestSnowflake`, key is `TS`, mode is `standard`. Click **Create**.
![bb-new-project](/static/blog/database-change-management-with-snowflake/bb-new-project.webp)

## Step 4 - Create a database in Snowflake via Bytebase

In Bytebase, a **Database** is the one created by `CREATE DATABASE xxx`. A database always belongs to a single **Project**. **Issue** represents a specific collaboration activity between Developer and DBA such as creating a database, altering a schema. It's similar to the issue concept in other issue management tools.

1. After the project is created. Follow the onboarding guide if it’s your first time using Bytebase. If not, click **New DB** on the top bar.
![bb-new-db](/static/blog/database-change-management-with-snowflake/bb-new-db.webp)

2. Fill the form with Name - `DB_DEMO_BB` (BB is short for Bytebase), Environment - `Test`, and Instance - `Snowflake instance`. Click **Create**.
![bb-create-db](/static/blog/database-change-management-with-snowflake/bb-create-db.webp)

> 3. It will create an issue “CREATE DATABASE ….” automatically. Because it’s for `Test` environment, the issue will run without waiting for your approval by default. Click **Resolve**, and the issue is Done. The database is created.

Reviewer: ```suggestion
3. Bytebase will create an issue “CREATE DATABASE ….” automatically. Because it’s for the `Test` environment, the issue will run without waiting for your approval by default. Click **Resolve**, and the issue is Done. The database is created.
```

![bb-go-home](/static/blog/database-change-management-with-snowflake/bb-go-home.webp)

> 4. Go back to the home page by clicking **Home** on the left sidebar. If it’s the first time you use Bytebase, it’ll show a celebration. On the home page, you can see the project, the database, and the issue you just resolved.

Reviewer: ```suggestion
4. Go back to the home page by clicking **Home** on the left sidebar. If it’s the first time you use Bytebase, a celebration mark will appear. On the home page, you can see the project, the database, and the issue you just resolved.
```

3. Input title, SQL, and Assignee, and click **Create**.
![bb-is-new-create-table](/static/blog/database-change-management-with-snowflake/bb-is-new-create-table.webp)

4.Bytebase will do some basic checks and then execute the SQL. Since it’s for `Test` environment, the issue is automatically approved by default. Click **Resolve issue**.
![bb-is-create-table-run](/static/blog/database-change-management-with-snowflake/bb-is-create-table-run.webp)

4. The issue status will become Done.
![bb-is-create-table-done](/static/blog/database-change-management-with-snowflake/bb-is-create-table-done.webp)

5. On the issue page, click **view migration**. You will see diff for each migration.
![bb-view-migration](/static/blog/database-change-management-with-snowflake/bb-view-migration.webp)

6. You can also go to **Migration History** under the project to view the full history. Or go into a specific database to view its history.
![bb-prj-mh](/static/blog/database-change-management-with-snowflake/bb-prj-mh.webp)
![bb-db-mh](/static/blog/database-change-management-with-snowflake/bb-db-mh.webp)

## Bonus Section - Drift Detect

> This section requires you to have **Team Plan** or **Enterprise Plan** (you can start 14 days trial directly in the product without credit card).

Reviewer: ```suggestion
[Drift Detection]/docs/anomaly-detection/drift-detection) requires the **Team Plan** or **Enterprise Plan** (you can start 14 days trial directly in the product without a credit card).
```

Now you can see the full migration history of DB_DEMO_BB. However, what is **Establish new baseline**? When should it be used?

By adopting Bytebase, we expect teams to use Bytebase exclusively for all schema changes. Meanwhile, if someone has made Snowflake schema change outside of Bytebase, obviously Bytebase won’t know it. And because Bytebase has recorded its own copy of schema, when Bytebase compares that with the live schema having that out-of-band schema change, it will notice a discrepancy and surface a schema drift anomaly. If that change is intended, then you should use baseline to reconcile the schema state again.

In this section, you’ll be guided through this process.

1. Go to Snowflake, and add a COLUMN there. Make sure the new column is added.
![sf-alter-add-age](/static/blog/database-change-management-with-snowflake/sf-alter-add-age.webp)

> 2. Wait for 10 mins or more. Go back to Bytebase, and you can find the Schema Drift on database DB_DEMO_BB

Reviewer: ```suggestion
2. Wait a few minutes before going back to Bytebase, and you will find the Schema Drift on database DB_DEMO_BB
```

Author: @milasuperstar Bytebase does the check every 10 mins.

Reviewer: can we say "Wait for 10 mins (as Bytebase does the check every 10 mins). Go back to Bytebase, and you can find the Schema Drift on database DB_DEMO_BB"

Author: Let's do it!

![bb-db-schema-drift](/static/blog/database-change-management-with-snowflake/bb-db-schema-drift.webp)

The Anomaly Center also surfaces the drift
![bb-ac-drift](/static/blog/database-change-management-with-snowflake/bb-ac-drift.webp)

3. Click **View diff**, you will see the exact drift.
![bb-view-drift](/static/blog/database-change-management-with-snowflake/bb-view-drift.webp)

4. Use baseline to reconcile the schema state from the live database schema. Go to DB_DEMO_BB > **Migration History** and click **Establish new baseline**.
![bb-db-establish-new-baseline](/static/blog/database-change-management-with-snowflake/bb-db-establish-new-baseline.webp)

5. It will create an issue. Click **Resolve** to make it done.
![bb-is-baseline-done](/static/blog/database-change-management-with-snowflake/bb-is-baseline-done.webp)

6. Go back to DB_DEMO_BB or Anomaly Center, and you will find the Drift is gone.
![bb-db-no-anomalies](/static/blog/database-change-management-with-snowflake/bb-db-no-anomalies.webp)
![bb-ac-no-anomaly](/static/blog/database-change-management-with-snowflake/bb-ac-no-anomaly.webp)

## Summary and Next

> Now you have connected snowflake with Bytebase, and tried out the UI workflow to do schema change. Bytebase will record the full migration history for you. With **Team or Enterprise Plan**, you can even have drift detection.

Reviewer: ```suggestion
Now you have connected Snowflake with Bytebase, and tried out the UI workflow to do schema change. Bytebase will record the full migration history for you. With **Team or Enterprise Plan**, you will have drift detection.
```

---
title: Database Change Management with Snowflake
author: Ningjing
published_at: 2022/12/22 21:15

> feature_image: /static/blog/database-change-management-with-snowflake/snowflake-db-management.webp

Author: What should the image look like? The 1.9.1 article has a list.

3. Input title, SQL, and Assignee, and click **Create**.
![bb-is-new-create-table](/static/blog/database-change-management-with-snowflake/bb-is-new-create-table.webp)

4.Bytebase will do some basic checks and then execute the SQL. Since it’s for `Test` environment, the issue is automatically approved by default. Click **Resolve issue**.
![bb-is-create-table-run](/static/blog/database-change-management-with-snowflake/bb-is-create-table-run.webp)

4. The issue status will become Done.
![bb-is-create-table-done](/static/blog/database-change-management-with-snowflake/bb-is-create-table-done.webp)

5. On the issue page, click **view migration**. You will see diff for each migration.
![bb-view-migration](/static/blog/database-change-management-with-snowflake/bb-view-migration.webp)

6. You can also go to **Migration History** under the project to view the full history. Or go into a specific database to view its history.
![bb-prj-mh](/static/blog/database-change-management-with-snowflake/bb-prj-mh.webp)
![bb-db-mh](/static/blog/database-change-management-with-snowflake/bb-db-mh.webp)

> ## Bonus Section - Drift Detect

Reviewer 2: ```suggestion
## Bonus Section - Schema Drift Detection
```

This section requires you to have **Team Plan** or **Enterprise Plan** (you can start 14 days trial directly in the product without credit card).

Now you can see the full migration history of DB_DEMO_BB. However, what is **Establish new baseline**? When should it be used?

By adopting Bytebase, we expect teams to use Bytebase exclusively for all schema changes. Meanwhile, if someone has made Snowflake schema change outside of Bytebase, obviously Bytebase won’t know it. And because Bytebase has recorded its own copy of schema, when Bytebase compares that with the live schema having that out-of-band schema change, it will notice a discrepancy and surface a schema drift anomaly. If that change is intended, then you should use baseline the schema state again to reconcile.

In this section, you’ll be guided through this process.

1. Go to Snowflake, and add a COLUMN there. Make sure the new column is added.
![sf-alter-add-age](/static/blog/database-change-management-with-snowflake/sf-alter-add-age.webp)

> 2. Wait for 10 mins or more. Go back to Bytebase, and you can find the Schema Drift on
> database DB_DEMO_BB

Reviewer 2: ```suggestion
2. Wait for 10 mins or more. Go back to Bytebase, and you can find the Schema Drift on database DB_DEMO_BB
```

![bb-db-schema-drift](/static/blog/database-change-management-with-snowflake/bb-db-schema-drift.webp)

> Anomaly Center

Reviewer 2: ```suggestion
The Anomaly Center also surfaces the drift
```

![bb-ac-drift](/static/blog/database-change-management-with-snowflake/bb-ac-drift.webp)

3. Click **View diff**, you will see the exact drift.
![bb-view-drift](/static/blog/database-change-management-with-snowflake/bb-view-drift.webp)

> 4. Go to DB_DEMO_BB > **Migration History** and click **Establish new baseline**.

Reviewer 2: ```suggestion
4. Use baseline to reconcile the schema state from the live database schema. Go to DB_DEMO_BB > **Migration History** and click **Establish new baseline**.
```

![bb-db-establish-new-baseline](/static/blog/database-change-management-with-snowflake/bb-db-establish-new-baseline.webp)

5. It will create an issue. Click **Resolve** to make it done.
![bb-is-baseline-done](/static/blog/database-change-management-with-snowflake/bb-is-baseline-done.webp)

6. Go back to DB_DEMO_BB or Anomaly Center, and you will find the Drift is gone.
![bb-db-no-anomalies](/static/blog/database-change-management-with-snowflake/bb-db-no-anomalies.webp)
![bb-ac-no-anomaly](/static/blog/database-change-management-with-snowflake/bb-ac-no-anomaly.webp)

## Summary and Next

> Now you have connected snowflake with Bytebase, and tried out the UI workflow to do schema change. Bytebase will record the full migration history for you. With **Team or Enterprise Plan**, you can even have drift detection.

Reviewer 2: ```suggestion
Now you have connected snowflake with Bytebase, and tried out the UI workflow to do schema change. Bytebase will record the full migration history for you. With **Team or Enterprise Plan**, you can even have schea drift detection.
```

> In the next article, you’ll try out GitOps workflow, which will store your Snowflake schema in GitHub and trigger the change upon committing the change to the repository, to bring your Snowflake change workflow to the next level of Database DevOps - Database as Code.

Reviewer 2: ```suggestion
In the next article, you’ll try out GitOps workflow, which will store your Snowflake schema in GitHub and trigger the change upon committing the change to the repository, to bring your Snowflake change workflow to the next level of Database DevOps - [Database as Code](/blog/database-as-code).
```
